compiler/luci/pass/src/FuseMulWithFullyConnectedPass.cpp

```diff
@@ -0,0 +1,218 @@
+                               const luci::CircleConst *multiplication)
+{
+  for (uint32_t i = 0; i < fused_node->size<loco::DataType::FLOAT32>(); i++)
+  {
+    fused_node->at<loco::DataType::FLOAT32>(i) *= multiplication->at<loco::DataType::FLOAT32>(0);
+  }
+}
+
+luci::CircleConst *gen_fused_weights(luci::CircleConst *weights,
+                                     const luci::CircleConst *multiplication)
+{
+  auto fused_weights = luci::clone(weights);
+  // Scalar multiplication:
+  if (is_scalar(multiplication))
+  {
+    update_with_scalar(fused_weights, multiplication);
+  }
+  // N-size multiplication:
+  else
+  {
+    // Go along channels, multiplication size is ensured to be compatible with channels.
+    auto count = fused_weights->dim(0).value();
+    auto size = fused_weights->dim(fused_weights->rank() - 1).value();
+    float val;
+    for (uint32_t c = 0; c < count; c++)
+    {
+      val = multiplication->at<loco::DataType::FLOAT32>(c);
+      for (uint32_t i = 0; i < size; i++)
+      {
+        fused_weights->at<loco::DataType::FLOAT32>(c * size + i) *= val;
+      }
+    }
+  }
+  return fused_weights;
+}
+
+luci::CircleConst *gen_fused_bias(luci::CircleConst *bias, const luci::CircleConst *multiplication)
+{
+  auto fused_bias = luci::clone(bias);
+  // Scalar multiplication:
+  if (is_scalar(multiplication))
+  {
+    update_with_scalar(fused_bias, multiplication);
+  }
+  // N-size multiplication:
+  else
+  {
+    // Go along channels, multiplication size is ensured to be compatible with channels.
+    for (uint32_t i = 0; i < fused_bias->size<loco::DataType::FLOAT32>(); i++)
+    {
+      fused_bias->at<loco::DataType::FLOAT32>(i) *= multiplication->at<loco::DataType::FLOAT32>(i);
+    }
+  }
+  return fused_bias;
+}
+
+/**
+ *  Fuse Mul to FullyConnected if the multiplied value is a channel(last dimension)-wise constant
+ *
+ *  BEFORE
+ *                |
+ *      [CircleFullyConnected]
+ *                |
+ *           [CircleMul]
+ *                |
+ *
+ *  AFTER
+ *                |
+ *       [CircleFullyConnected]   [CircleMul] (dead)
+ *                |
+ *
+ */
+bool fuse_mul_with_fc(luci::CircleFullyConnected *fc)
+{
+  // Sanity check:
+  RETURN_FALSE_UNLESS(fc);
+  // Allow only FLOAT32 data type:
+  RETURN_FALSE_UNLESS(fc->dtype() == loco::DataType::FLOAT32);
+  // Allow only without activation functions as values are going to
+  // be multiplied before activation function.
+  RETURN_FALSE_UNLESS(fc->fusedActivationFunction() == luci::FusedActFunc::NONE);
+  // Check for weights being Constant:
+  auto weights = dynamic_cast<luci::CircleConst *>(fc->weights());
+  RETURN_FALSE_UNLESS(weights);
+  // Get Mul node:
+  auto fc_output = loco::succs(fc);
+  // Make sure that FullyConnected has only one child:
+  RETURN_FALSE_UNLESS(fc_output.size() == 1);
+  auto mul = dynamic_cast<luci::CircleMul *>(*fc_output.begin());
+  RETURN_FALSE_UNLESS(mul);
```

Comment:
We usually go upward(to input).
For this case, begin with `Mul` and check one of the input is `FC`.
Can you please revise in this check order?

Reply:
When you traverse from `Mul` to `FC`, `FC` input of `Mul` should be created and replaced to new `FC` as exiting `FC` may have multiple successors.

Current check only works when there exist single connection of `FC`-`Mul`.

Reply:
I see approach similar to mine on other pass which base on `FC` as well (with `Add` instead of `Mul`):
https://github.com/Samsung/ONE/blob/719a6f7d8b3873b3bf06bd4c17ea7fe77288d7cd/compiler/luci/pass/src/FuseAddWithFullyConnectedPass.cpp#L43
https://github.com/Samsung/ONE/blob/719a6f7d8b3873b3bf06bd4c17ea7fe77288d7cd/compiler/luci/pass/src/FuseAddWithFullyConnectedPass.cpp#L58-L61

I want to make sure I understand the motivation here. Here is my understanding of a graph and the approach to pass you described:
![image](https://github.com/user-attachments/assets/ed8818cc-b24c-4f83-9c16-cc4db6ee721a)

In case of such pattern of `FC`, I understand the fusion of `Mul` into `FC` to eliminate node and reduce overall number of operations. However, if `FC` has more operations connected to it's output, the pass is simply replacing one op for another -- which from "math view" is resulting in the same count of operations.
Is that a matter of existing optimizations in targeted HW and/or more performant kernel implementations in ONE that FC is preferred here in place of `Mul`?

Reply:
> Is that a matter of existing optimizations in targeted HW and/or more performant kernel implementations in ONE that FC is preferred here in place of Mul?

No.

> is that something that can be considered instead of adding multiple FC?

I cannot understand your question point.

Reply:
`FuseAddWithFullyConnected` fuses add with FC only when FC has a single successor (add). That is an intended behavior. If FC has multiple successors, FC layers are duplicated (as @jiwaszki said), which tends to increase model size and degrade model performance. I think this pass can follow the same strategy with `FuseAddWithFullyConnected`.

So, +1 for the current implementation.

Reply:
I DISAGREE. I DO NOT WANT MULTIPLE WAYS OF SEARCHING.
SOME DAY FuseAddWithFullyConnected SHOULD BE FIXED TOO.

Reply:
@seanshpark How about the changes in https://github.com/Samsung/ONE/pull/13607/commits/79a2213d063bbf85e0209a5ec06dc5030cc7c80c ? It merges both ideas into one.
First, it follows the bottom-up approach, starting from `mul`:
https://github.com/Samsung/ONE/blob/79a2213d063bbf85e0209a5ec06dc5030cc7c80c/compiler/luci/pass/src/FuseMulWithFullyConnectedPass.cpp#L110
https://github.com/Samsung/ONE/blob/79a2213d063bbf85e0209a5ec06dc5030cc7c80c/compiler/luci/pass/src/FuseMulWithFullyConnectedPass.cpp#L116-L120

Second, still has the intended behavior of fusing only when there is only one successor to `fc`:
https://github.com/Samsung/ONE/blob/79a2213d063bbf85e0209a5ec06dc5030cc7c80c/compiler/luci/pass/src/FuseMulWithFullyConnectedPass.cpp#L121-L122

@jinevening please also review the new commit when you have time.

Reply:
It looks good to me. Please address @seanshpark 's comments.